pkg/proxy/iptables/proxier.go

```diff
@@ -656,6 +664,36 @@ func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
 	proxier.deleteEndpointConnections(staleConnections)
 }
 
+// needToUpdate returns if the input endpoints is different from the last batch
+// This is calculated by comparing the checksum of EndpointSubsets.
+func (proxier *Proxier) needToUpdate(allEndpoints []api.Endpoints) bool {
+	checksumMap := make(map[string]uint32)
+	hash := fnv.New32a()
+
+	for _, ep := range allEndpoints {
+		hashutil.DeepHashObject(hash, ep.Subsets)
+		checksumMap[fmt.Sprintf("%s/%s", ep.Namespace, ep.Name)] = hash.Sum32()
+	}
+
+	needToUpdate := false
+	if len(checksumMap) == len(proxier.endpointChecksums) {
+		for ep, checksum := range checksumMap {
+			if checksum != proxier.endpointChecksums[ep] {
+				needToUpdate = true
+				break
+			}
+		}
+	} else {
+		needToUpdate = true
+	}
+
+	if needToUpdate {
+		proxier.endpointChecksums = checksumMap
```

Comment:
One way to refactor thread safety and get/set is (doesn't matter if thread safety is not an issue)
```go
type endpointHasher struct {
  oldEndpoints 
}

// setEndpoints sets the internal field
// not threadsafe
func (e *endpointHasher) setEndpoints(newEndpoints) {}

// isEqual compares a hash of the internal field with the hash of the given newEndpoints
// not threadsafe
func (e *endpointHasher) isEqual(newEndpoints) {}

// ShouldUpdate returns true if the newEndpoints require an update.
// Also stores newEndpoints in the internal field.
func (e *endpointHasher) ShouldUpdate(newEndpoints) {
   e.lock.Lock()
   defer e.lock.Unlock()

   if e.isEqual() {
    return true
  }

  e.setEndpoints(newEndpoints)
  return false
}
```